### PR TITLE
feat(gRPC): adding AvailabilityScore to gRPC

### LINKED ISSUE #907: Adding Availability Score to gRPC
We need to add Availability Score to gRPC module. every where we return validators info, we need to add Availability Score parameter to it.

### PR DESCRIPTION
**Pull request by izikdepth**
- Fixes #907

### CODE REVIEW

File: www/grpc/blockchain.go [hunk 1]
```diff
@@ -233,7 +234,7 @@ func (s *blockchainServer) GetPublicKey(_ context.Context,
 	return &pactus.GetPublicKeyResponse{PublicKey: publicKey.String()}, nil
 }
 
-func validatorToProto(val *validator.Validator) *pactus.ValidatorInfo {
+func validatorToProto(val *validator.Validator, state state.Facade) *pactus.ValidatorInfo {
```
Comment (kehiy) [inline www/grpc/blockchain.go:237]:
```suggestion
func validatorToProto(val *validator.Validator, sf state.Facade) *pactus.ValidatorInfo {
```

File: www/grpc/blockchain.go [hunk 2]
```diff
@@ -245,6 +246,7 @@ func validatorToProto(val *validator.Validator) *pactus.ValidatorInfo {
 		LastBondingHeight:   val.LastBondingHeight(),
 		LastSortitionHeight: val.LastSortitionHeight(),
 		UnbondingHeight:     val.UnbondingHeight(),
+		AvailabilityScore:   float32(state.AvailabilityScore(val.Number())),
```
Comment (kehiy) [inline www/grpc/blockchain.go:249]:
```suggestion
		AvailabilityScore:   float32(sf.AvailabilityScore(val.Number())),
```

**Review by kehiy — CHANGES_REQUESTED** [review]

**Review by kehiy — APPROVED** [review]

**Comment by kehiy** [comment]
@izikdepth Thanks!

File: www/grpc/blockchain.go [hunk 3]
```diff
@@ -233,7 +234,7 @@ func (s *blockchainServer) GetPublicKey(_ context.Context,
 	return &pactus.GetPublicKeyResponse{PublicKey: publicKey.String()}, nil
 }
 
-func validatorToProto(val *validator.Validator) *pactus.ValidatorInfo {
+func validatorToProto(val *validator.Validator, sf state.Facade) *pactus.ValidatorInfo {
```
Comment (b00f) [inline www/grpc/blockchain.go:237]: @izikdepth 

you can make it a member function, like:
```
func (s *blockchainServer) validatorToProto(val *validator.Validator) *pactus.ValidatorInfo {
...
}
```

Then you can access to `state` like `s.state`

Comment (b00f) [reply www/grpc/blockchain.go:237]: Please do the same for: `accountToProto` and `voteToProto`

**Review by b00f — CHANGES_REQUESTED** [review]

File: www/grpc/blockchain.go (hunk 3 above)
Comment (izikdepth) [reply www/grpc/blockchain.go:237]: is this correct?

```func (s *blockchainServer) validatorToProto(val *validator.Validator) *pactus.ValidatorInfo {
	data, _ := val.Bytes()
	return &pactus.ValidatorInfo{
		Hash:                val.Hash().Bytes(),
		Data:                data,
		PublicKey:           val.PublicKey().String(),
		Address:             val.Address().String(),
		Number:              val.Number(),
		Stake:               val.Stake(),
		LastBondingHeight:   val.LastBondingHeight(),
		LastSortitionHeight: val.LastSortitionHeight(),
		UnbondingHeight:     val.UnbondingHeight(),
		AvailabilityScore:   float32(s.state.AvailabilityScore(val.Number())),
	}
}

func (s *blockchainServer) accountToProto(addr crypto.Address, acc *account.Account) *pactus.AccountInfo {
	data, _ := acc.Bytes()
	return &pactus.AccountInfo{
		Hash:    acc.Hash().Bytes(),
		Data:    data,
		Number:  acc.Number(),
		Balance: acc.Balance(),
		Address: addr.String(),
	}
}

func (s *blockchainServer) voteToProto(v *vote.Vote) *pactus.VoteInfo {
	cpRound := int32(0)
	cpValue := int32(0)
	if v.IsCPVote() {
		cpRound = int32(v.CPRound())
		cpValue = int32(v.CPValue())
	}

	return &pactus.VoteInfo{
		Type:      pactus.VoteType(v.Type()),
		Voter:     v.Signer().String(),
		BlockHash: v.BlockHash().Bytes(),
		Round:     int32(v.Round()),
		CpRound:   cpRound,
		CpValue:   cpValue,
	}
}```

Comment (kehiy) [reply www/grpc/blockchain.go:237]: @izikdepth something like that:
```go
func (bs *blockchainServer) validatorToProto(val *validator.Validator) *pactus.ValidatorInfo {
	data, _ := val.Bytes()
	return &pactus.ValidatorInfo{
		Hash:                val.Hash().Bytes(),
		Data:                data,
		PublicKey:           val.PublicKey().String(),
		Address:             val.Address().String(),
		Number:              val.Number(),
		Stake:               val.Stake(),
		LastBondingHeight:   val.LastBondingHeight(),
		LastSortitionHeight: val.LastSortitionHeight(),
		UnbondingHeight:     val.UnbondingHeight(),
                AvailabilityScore:   float32(bs.state.AvailabilityScore(val.Number())),
	}
}

```

**Comment by kehiy** [comment]
@b00f I think there is no need for account and vote to proto function to be methods of blockchain server!? they don't use any part of it.

**Comment by b00f** [comment]
@kehiy let's keep all the same. You are write, but that way we have harmony in our functions.

File: www/grpc/blockchain.go [hunk 4]
```diff
@@ -233,7 +233,7 @@ func (s *blockchainServer) GetPublicKey(_ context.Context,
 	return &pactus.GetPublicKeyResponse{PublicKey: publicKey.String()}, nil
 }
 
-func validatorToProto(val *validator.Validator) *pactus.ValidatorInfo {
+func (bs *blockchainServer) validatorToProto(val *validator.Validator) *pactus.ValidatorInfo {
```
Comment (kehiy) [inline www/grpc/blockchain.go:236]:
```suggestion
func (s *blockchainServer) validatorToProto(val *validator.Validator) *pactus.ValidatorInfo {
```

File: www/grpc/blockchain.go [hunk 5]
```diff
@@ -245,10 +245,11 @@ func validatorToProto(val *validator.Validator) *pactus.ValidatorInfo {
 		LastBondingHeight:   val.LastBondingHeight(),
 		LastSortitionHeight: val.LastSortitionHeight(),
 		UnbondingHeight:     val.UnbondingHeight(),
+		AvailabilityScore:   float32(bs.state.AvailabilityScore(val.Number())),
```
Comment (kehiy) [inline www/grpc/blockchain.go:248]:
```suggestion
		AvailabilityScore:   float32(s.state.AvailabilityScore(val.Number())),
```

File: www/grpc/blockchain.go [hunk 6]
```diff
@@ -245,10 +245,11 @@ func validatorToProto(val *validator.Validator) *pactus.ValidatorInfo {
...
 	}
 }
 
-func accountToProto(addr crypto.Address, acc *account.Account) *pactus.AccountInfo {
+func (bs *blockchainServer) accountToProto(addr crypto.Address, acc *account.Account) *pactus.AccountInfo {
```
Comment (kehiy) [inline www/grpc/blockchain.go:252]:
```suggestion
func (s *blockchainServer) accountToProto(addr crypto.Address, acc *account.Account) *pactus.AccountInfo {
```

File: www/grpc/blockchain.go [hunk 7]
```diff
@@ -259,7 +260,7 @@ func accountToProto(addr crypto.Address, acc *account.Account) *pactus.AccountIn
 	}
 }
 
-func voteToProto(v *vote.Vote) *pactus.VoteInfo {
+func (bs *blockchainServer) voteToProto(v *vote.Vote) *pactus.VoteInfo {
```
Comment (kehiy) [inline www/grpc/blockchain.go:263]:
```suggestion
func (s *blockchainServer) voteToProto(v *vote.Vote) *pactus.VoteInfo {
```

**Review by kehiy — CHANGES_REQUESTED** [review]

**Review by kehiy — APPROVED** [review]

**Comment by kehiy** [comment]
@b00f we can merge it.

File: www/grpc/proto/blockchain.proto [hunk 8]
```diff
@@ -90,6 +90,7 @@ message ValidatorInfo {
   uint32 last_sortition_height = 7;
   uint32 unbonding_height = 8;
   string address = 9;
+  float availability_score = 10;
```
Comment (b00f) [inline www/grpc/proto/blockchain.proto:93]:
```suggestion
  double availability_score = 10;
```

File: www/grpc/blockchain.go [hunk 9]
```diff
@@ -245,10 +245,11 @@ func validatorToProto(val *validator.Validator) *pactus.ValidatorInfo {
 		LastBondingHeight:   val.LastBondingHeight(),
 		LastSortitionHeight: val.LastSortitionHeight(),
 		UnbondingHeight:     val.UnbondingHeight(),
+		AvailabilityScore:   bs.state.AvailabilityScore(val.Number()),
```
Comment (b00f) [inline www/grpc/blockchain.go:248]:
```suggestion
		AvailabilityScore:   s.state.AvailabilityScore(val.Number()),
```

**Comment by b00f** [comment]
@kehiy Please merge it is tests are passed. Then we can integrate it with our Bot

**Comment by kehiy** [comment]
@izikdepth run `make proto` then push your changes.
proto files dose not re generated.

**Review by b00f — APPROVED** [review]

**Comment by b00f** [comment]
@izikdepth Thanks

**Pull request merged**